File: vfr-green-map/src/resources/cached-provider.ts

import { PriorityArray } from '../utility/priority-array'

export abstract class CachedProvider {
    private static readonly MAX_ACTIVE_REQUESTS = 2;
    private cache: Record<string, CachedProviderRequest>; //Need to add ageoff, causing memory leak
    private numActiveRequests: number;
    private requestQueue: PriorityArray<[string, CachedProviderRequest]>;
    private requestQueueKeys: Record<string,number>;    //key, priority
    private requestDelayMilliseconds: number;
    private lastQueueAddition: Date;
    private processQueuePending: boolean;

    constructor(requestDelayMilliseconds: number = 0) {
        this.cache = {};
        this.numActiveRequests = 0;
        this.requestQueue = new PriorityArray<[string, CachedProviderRequest]>();
        this.requestQueueKeys = {};
        this.requestDelayMilliseconds = requestDelayMilliseconds;
        this.lastQueueAddition = new Date();
        this.processQueuePending = false;
    }

    public isLoading(): boolean {
        return this.numActiveRequests > 0 || this.requestQueue.size() > 0;
    }

    public clearQueue(): void {
        this.requestQueue.clear();
        this.requestQueueKeys = {};
    }

    public getExistingRequest(key: string): CachedProviderRequest | undefined {
        if ( key in this.cache ) {
            return this.cache[key];
        }
        if ( key in this.requestQueueKeys ) {
            let priority = this.requestQueueKeys[key];
            let priorityList = this.requestQueue.getPriorityList(priority);
            for ( let e of priorityList ) {
                if ( e[0] == key ) {
                    return e[1];
                }
            }
        }
    }

    /**
     * Only to be called by CachedProviderRequest class to signify a downoad has completed.
     */
    public completeRequest(): void {
        this.numActiveRequests--;
        this.processQueue();
    }

    protected addRequestToQueue(key: string, request: CachedProviderRequest): void {
        if (!(key in this.requestQueueKeys)) {
            this.requestQueueKeys[key] = request.getPriority();
            this.requestQueue.add([key, request], request.getPriority());
            this.lastQueueAddition = new Date();
            this.processQueue();
        }
    }

    protected addRequestToCache(key: string, request: CachedProviderRequest): void {
        this.cache[key] = request;
    }

    private processQueue(): void {
        if ( this.processQueuePending ) {
            return;
        }
        let now = new Date();
        let timeToWait = this.requestDelayMilliseconds - (now.getTime() - this.lastQueueAddition.getTime())
        if ( timeToWait <= 0 ) {
            while ( this.numActiveRequests < CachedProvider.MAX_ACTIVE_REQUESTS && this.requestQueue.size() > 0 ) {
                let request = this.requestQueue.pop();
                if ( request === undefined ) {
                    break;
                }
                delete this.requestQueueKeys[request[0]];
                this.addRequestToCache(request[0], request[1]);
                request[1].sendRequest();
                this.numActiveRequests++;
            }
        } else {
            this.processQueuePending = true;
            setTimeout(() => {
                this.processQueuePending = false;
                this.processQueue();
            }, timeToWait);
        }
    }
};

export abstract class CachedProviderRequest {
    private loaded: boolean;
    private inError: boolean;
    private provider: CachedProvider;
    private priority: number;

    constructor(provider: CachedProvider, priority: number) {
        this.loaded = false;
        this.inError = false;
        this.provider = provider;
        this.priority = priority;
    }

    abstract sendRequest(): void;

    abstract broadcastData(immediate: boolean): void;

    public getPriority(): number {
        return this.priority;
    }

    public isLoaded(): boolean {
        return this.loaded && !this.inError;
    }

    public isInError(): boolean {
        return this.inError;
    }

    protected setLoaded(loaded: boolean = true): void {
        this.loaded = loaded
        this.inError = false;
    }

    protected setError(): void {
        this.inError = true;
        this.loaded = false;
    }

    protected completeRequest(isSuccess: boolean): void {
        this.provider.completeRequest();
        if ( isSuccess ) {
            this.setLoaded();
            this.broadcastData(false);
        } else {
            this.setError();
        }
    }
}

File: vfr-green-app/app/src/main/java/com/blueskycharts/app/map/resources/CachedProvider.kt
package com.blueskycharts.app.map.resources

import com.blueskycharts.app.utility.PriorityArray
import com.blueskycharts.app.map.view.Map
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import kotlin.concurrent.timerTask

abstract class CachedProvider( private val map: Map, private val requestDelayMilliseconds: Int = 0 ) {
    private val maxActiveRequests = 2
    private val cache = Hashtable<String, CachedProviderRequest>(); //Need to add ageoff, causing memory leak
    private var numActiveRequests = AtomicInteger(0)
    private var numAwaitingQueueAddition = AtomicInteger(0)
    private val requestQueue: PriorityArray<Pair<String, CachedProviderRequest>> = PriorityArray();
    private var requestQueueKeys = Hashtable<String, Int>();    //key, priority
    private var lastQueueAddition = Date();
    private var processQueuePending = false;
    private val processQueueTimer = Timer(false);
    private val queueMutex: Mutex = Mutex()

    fun isLoading(): Boolean {
        return this.numActiveRequests.get() > 0 || this.requestQueue.size() > 0 || numAwaitingQueueAddition.get() > 0
    }

    fun clearQueue() {
        GlobalScope.launch {    //ok1
            queueMutex.withLock {
                this@CachedProvider.requestQueue.clear();
                this@CachedProvider.requestQueueKeys = Hashtable();
            }
        }
    }

    fun getCachedItem(key: String) :CachedProviderRequest? {
        return try {
            this.cache[key]
        } catch ( e: Throwable ) {
            null
        }
    }

    suspend fun getExistingRequest(key: String): CachedProviderRequest? {
        queueMutex.withLock {
            if (key in this.cache.keys) {
                return this.cache[key];
            }
            if (key in this.requestQueueKeys.keys) {
                val priority = this.requestQueueKeys[key] ?: return null;
                val priorityList = this.requestQueue.getPriorityList(priority);
                for (e in priorityList) {
                    if (e.first == key) {
                        return e.second;
                    }
                }
            }
            return null;
        }
    }

    /**
     * Only to be called by CachedProviderRequest class to signify a downoad has completed.
     */
    fun completeRequest() {
        this.numActiveRequests.getAndDecrement();
        this.processQueue();
        this.map.requestRedraw()
    }

    protected fun incrementAwaitingQueueAddition() {
        numAwaitingQueueAddition.getAndIncrement()
    }

    protected fun decrementAwaitingQueueAddition() {
        numAwaitingQueueAddition.getAndDecrement()
    }

    protected suspend fun addDataToCache(key: String, request: CachedProviderRequest) {
        queueMutex.withLock {
            cache[key] = request
        }
    }

    protected fun addRequestToQueue(key: String, request: CachedProviderRequest) {
        //make it so a loaded request adds to the cache if it's receive date is newer than what's there or there's nothing there.
        if (key in this.requestQueueKeys) {
            return
        }
        incrementAwaitingQueueAddition()
        GlobalScope.launch {    //ok1
            queueMutex.withLock {
                try {
                    if (key in this@CachedProvider.requestQueueKeys.keys) {
                        return@launch
                    }
                    try {
                        if (this@CachedProvider.cache.containsKey(key)) {
                            val cacheItem: CachedProviderRequest? = this@CachedProvider.cache[key]
                            if (cacheItem != null && !cacheItem.inError && !(cacheItem.loaded && cacheItem.expired)) {
                                return@launch
                            }
                        }
                    } catch (e: Throwable) {
                    }

                    this@CachedProvider.requestQueueKeys[key] = request.priority
                    this@CachedProvider.requestQueue.add(Pair(key, request), request.priority)
                } finally {
                    decrementAwaitingQueueAddition()
                }
            }
            this@CachedProvider.lastQueueAddition = Date();
            this@CachedProvider.processQueue();
        }
    }

    private fun processQueue() {
        if (this.processQueuePending) {
            return
        }
        val now = Date();
        val timeToWait = this.requestDelayMilliseconds - (now.time - this.lastQueueAddition.time)
        if (timeToWait <= 0) {
            GlobalScope.launch {
                queueMutex.withLock {
                    while (this@CachedProvider.numActiveRequests.get() < maxActiveRequests && this@CachedProvider.requestQueue.size() > 0) {
                        val request = this@CachedProvider.requestQueue.pop()
                        if (request != null) {
                            this@CachedProvider.numActiveRequests.getAndIncrement();
                            this@CachedProvider.cache[request.first] = request.second;
                            request.second.sendRequest();
                            this@CachedProvider.requestQueueKeys.remove(request.first);
                        }
                    }
                }
            }
        } else {
            this.processQueuePending = true;
            val task: TimerTask = timerTask {
                processQueuePending = false;
                processQueue();
            }
            processQueueTimer.schedule(task, timeToWait)
        }
    }
};
